### PR TITLE
feat: add bank account OOP lab

### PR DESCRIPTION
**Pull request by zairahira**
Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/how-to-open-a-pull-request/).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes https://github.com/freeCodeCamp/CurriculumExpansion/pull/585/

Modified the original prototype by replacing `console.log` statements with `return` statements.

### CODE REVIEW

File: curriculum/challenges/english/25-front-end-development/lab-bank-account-manager/6718d2d59337c822ecb697ff.md [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+dashedName: build-a-bank-account-management-program
+---
+
+# --description--
+
+**Objective:** Fulfill the user stories below and get all the tests to pass to complete the lab. 
+
+**User Stories:**
+
+1. You should define a class named `BankAccount` with a constructor that takes `balance` as parameter. The constructor should set the default `balance` to `0`and initialize an empty array named `transactions` to store transaction records as objects.
```
Comment (Sembauke) [inline curriculum/challenges/english/25-front-end-development/lab-bank-account-manager/6718d2d59337c822ecb697ff.md:14]:
```suggestion
1. You should define a class named `BankAccount` with a constructor that takes `balance` as a parameter. The constructor should set the `balance` to `0` if no value is provided and initialize an empty array named `transactions` to store transaction records as objects.
```

File: curriculum/challenges/english/25-front-end-development/lab-bank-account-manager/6718d2d59337c822ecb697ff.md [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+
+# --description--
+
+**Objective:** Fulfill the user stories below and get all the tests to pass to complete the lab. 
+
+**User Stories:**
+
+1. You should define a class named `BankAccount` with a constructor that takes `balance` as parameter. The constructor should set the default `balance` to `0`and initialize an empty array named `transactions` to store transaction records as objects. 
+
+2. The `transactions` object defined inside the `BankAccount` constructor should be able to take two properties: `type` and `amount`. The `type` property should be either `deposit` or `withdraw`, and the `amount` property should be the amount deposited or withdrawn.
```
Comment (Sembauke) [inline curriculum/challenges/english/25-front-end-development/lab-bank-account-manager/6718d2d59337c822ecb697ff.md:16]:
```suggestion
2. Each transaction stored in the `transactions` array should be an object with two properties: `type` and `amount`. The `type` property should be either `deposit` or `withdraw`, and the `amount` property should be the amount deposited or withdrawn.
```

File: curriculum/challenges/english/25-front-end-development/lab-bank-account-manager/6718d2d59337c822ecb697ff.md [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+
+3. Inside the `BankAccount` class, You should define a method named `updateBalance` that updates the balance according to the type of transaction. If the transaction type is `deposit`, it should add the amount to the balance. If the transaction type is `withdraw`, it should subtract the amount from the balance.
```
Comment (Sembauke) [inline curriculum/challenges/english/25-front-end-development/lab-bank-account-manager/6718d2d59337c822ecb697ff.md:18]:
```suggestion
3. Inside the `BankAccount` class, you should define a method named `updateBalance` that updates the balance according to the type of transaction. If the transaction type is `deposit`, it should add the amount to the balance. If the transaction type is `withdraw`, it should subtract the amount from the balance.
```

File: curriculum/challenges/english/25-front-end-development/lab-bank-account-manager/6718d2d59337c822ecb697ff.md (hunk 2 above)
Comment (Sembauke) [reply curriculum/challenges/english/25-front-end-development/lab-bank-account-manager/6718d2d59337c822ecb697ff.md:16]: I think transactions is an array and not an object?

File: curriculum/challenges/english/25-front-end-development/lab-bank-account-manager/6718d2d59337c822ecb697ff.md [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+6. You should define a method named `withdraw` that takes an amount as a parameter. This method should update the current balance according to withdrawals. When the amount to be withdrawn is greater than `0` and less than or equal to the current balance, it should:
+
+   - Push a new object to the transactions array with a type of withdraw and the amount withdrawn.
+   - Update the balance.
+   - Return `"Successfully withdrew $[amount]. New balance: $[balance]"`. 
+
+7. In the `withdraw` method, when the amount to be withdrawn is less than or equal to `0` or greater than the current balance, it should return `"Insufficient balance or invalid amount."`.
+
+
+8. You should define a method named `checkBalance` that returns the current balance in the format `"Current balance: $[balance]"`.
```
Comment (Sembauke) [inline curriculum/challenges/english/25-front-end-development/lab-bank-account-manager/6718d2d59337c822ecb697ff.md:37]:
```suggestion
8. You should define a method named `checkBalance` that returns the current balance in the format `"Current balance: ${balance}"`.
```

**Review by Sembauke — CHANGES_REQUESTED** [review]

File: curriculum/challenges/english/25-front-end-development/lab-bank-account-manager/6718d2d59337c822ecb697ff.md (hunk 4 above)
Comment (zairahira) [reply curriculum/challenges/english/25-front-end-development/lab-bank-account-manager/6718d2d59337c822ecb697ff.md:37]: I have used square brackets as placeholders here and not as template literals.

File: curriculum/challenges/english/25-front-end-development/lab-bank-account-manager/6718d2d59337c822ecb697ff.md (hunk 1 above)
Comment (zairahira) [reply curriculum/challenges/english/25-front-end-development/lab-bank-account-manager/6718d2d59337c822ecb697ff.md:14]: The constructor would always set balance to `0` and doesn't accept other values.

File: curriculum/challenges/english/25-front-end-development/lab-bank-account-manager/6718d2d59337c822ecb697ff.md (hunk 4 above)
Comment (ilenia-magoni) [reply curriculum/challenges/english/25-front-end-development/lab-bank-account-manager/6718d2d59337c822ecb697ff.md:37]:
```suggestion
8. You should define a method named `checkBalance` that returns the current balance in the format `"Current balance: $<balance>"`.
```

I have seen in the curriculum most often the angular brackets to signal placeholder

Comment (Sembauke) [reply curriculum/challenges/english/25-front-end-development/lab-bank-account-manager/6718d2d59337c822ecb697ff.md:37]: I am a bit confused but let's go with the optimal option @zairahira

**Review by Sembauke — APPROVED** [review]

File: curriculum/challenges/english/25-front-end-development/lab-bank-account-manager/6718d2d59337c822ecb697ff.md [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+
+**Objective:** Fulfill the user stories below and get all the tests to pass to complete the lab. 
+
+**User Stories:**
+
+1. You should define a class named `BankAccount` with a constructor that takes `balance` as parameter. The constructor should set the default `balance` to `0`and initialize an empty array named `transactions` to store transaction records as objects. 
+
+2. Each transaction stored in the `transactions` array should be an object with two properties: `type` and `amount`. The `type` property should be either `deposit` or `withdraw`, and the `amount` property should be the amount deposited or withdrawn.
+
+3. Inside the `BankAccount` class, you should define a method named `updateBalance` that updates the balance according to the type of transaction. If the transaction type is `deposit`, it should add the amount to the balance. If the transaction type is `withdraw`, it should subtract the amount from the balance.
```
Comment (ilenia-magoni) [inline curriculum/challenges/english/25-front-end-development/lab-bank-account-manager/6718d2d59337c822ecb697ff.md:18]:
```suggestion
3. Inside the `BankAccount` class, you should define a method named `updateBalance` that updates the balance using the `transactions` list, according to the type of transaction. If the transaction type is `deposit`, it should add the amount to the balance. If the transaction type is `withdraw`, it should subtract the amount from the balance.
```

I was really confused on what this method should do, there was no mention on using the list of all transactions to calculate the balance.

**Review by majestic-owl448 — COMMENTED** [review]
What do you think of adding a test that check if `this.transactions` contains the right objects?

**Comment by moT01** [comment]
A PR has been merged that renamed the certification from "Front End" to "Full Stack". This PR may need an update and a few changes made: [see comment](https://github.com/freeCodeCamp/freeCodeCamp/pull/56694#issuecomment-2444473378)

File: client/src/pages/learn/front-end-development/lab-bank-account-manager/index.md [hunk 6]
```diff
@@ -0,0 +1,9 @@
+---
+title: Introduction to the Build a Bank Account Management Program
+block: lab-bank-account-manager
+superBlock: front-end-development
```
Comment (ilenia-magoni) [inline client/src/pages/learn/front-end-development/lab-bank-account-manager/index.md:4]:
```suggestion
superBlock: full-stack-developer
```

File: curriculum/challenges/_meta/lab-bank-account-manager/meta.json [hunk 7]
```diff
@@ -0,0 +1,12 @@
+{
+  "name": "Build a Bank Account Management Program",
+  "isUpcomingChange": true,
+  "usesMultifileEditor": true,
+  "dashedName": "lab-bank-account-manager",
+  "order": 238,
+  "superBlock": "front-end-development",
```
Comment (ilenia-magoni) [inline curriculum/challenges/_meta/lab-bank-account-manager/meta.json:7]:
```suggestion
  "superBlock": "full-stack-developer",
```

**Comment by moT01** [comment]
A PR has been merged that moves the block order from the `meta.json` file to the `full-stack.json` file.](https://github.com/freeCodeCamp/freeCodeCamp/pull/56795) This PR will need an update, and the order will need to be removed from the `meta.json` file and moved to the new file.

File: curriculum/challenges/english/25-front-end-development/lab-bank-account-manager/6718d2d59337c822ecb697ff.md (hunk 5 above)
Comment (zairahira) [reply curriculum/challenges/english/25-front-end-development/lab-bank-account-manager/6718d2d59337c822ecb697ff.md:18]: The transactions is not a list, but there is a `transaction` array that has 2 types of transactions: `deposit` and `withdraw`.

File: curriculum/challenges/english/25-front-end-development/lab-bank-account-manager/6718d2d59337c822ecb697ff.md [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+```js
+
+```
+
+# --solutions--
+
+```js
+class BankAccount {
+    constructor(balance = 0) {
+      this.balance = balance;
```
[the remark this answers is not in the document]
Comment (zairahira) [reply curriculum/challenges/english/25-front-end-development/lab-bank-account-manager/6718d2d59337c822ecb697ff.md:217]: Makes sense 👍️

File: curriculum/challenges/english/25-front-end-development/lab-bank-account-manager/6718d2d59337c822ecb697ff.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (zairahira) [reply curriculum/challenges/english/25-front-end-development/lab-bank-account-manager/6718d2d59337c822ecb697ff.md:18]: This is being tested inside `withdraw()` and `deposit()`. Should I still write separate tests?

File: curriculum/challenges/english/25-front-end-development/lab-bank-account-manager/6718d2d59337c822ecb697ff.md [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+
+**User Stories:**
+
+1. You should define a class named `BankAccount` with a constructor. The constructor should set the default `balance` to `0` and initialize an empty array named `transactions` to store transaction records as objects. 
+
+2. Each transaction stored in the `transactions` array should be an object with two properties: `type` and `amount`. The `type` property should be either `deposit` or `withdraw`, and the `amount` property should be the amount deposited or withdrawn.
+
+3. Inside the `BankAccount` class, you should define a method named `updateBalance` that updates the balance according to the type of transaction. If the transaction type is `deposit`, it should add the amount to the balance. If the transaction type is `withdraw`, it should subtract the amount from the balance.
+
+4. You should define a method named `deposit` that takes the deposit amount as a parameter. This method should update the current balance according to deposits. When the deposit amount is greater than `0`, it should:
```
Comment (moT01) [inline curriculum/challenges/english/25-front-end-development/lab-bank-account-manager/6718d2d59337c822ecb697ff.md:20]:
```suggestion
4. You should define a method named `deposit` that takes the deposit amount as a parameter. When the deposit amount is greater than `0`, it should:
```

This part seems redundant. It says to update the balance below.

File: curriculum/challenges/english/25-front-end-development/lab-bank-account-manager/6718d2d59337c822ecb697ff.md [hunk 10]
```diff
@@ -0,0 +1,275 @@
...
+
+   - Push a new object to the transactions array with a type of deposit and the amount deposited.
```
Comment (moT01) [inline curriculum/challenges/english/25-front-end-development/lab-bank-account-manager/6718d2d59337c822ecb697ff.md:22]:
```suggestion
   - Push a new object to the transactions array with a type of `deposit` and the amount deposited.
```

File: curriculum/challenges/english/25-front-end-development/lab-bank-account-manager/6718d2d59337c822ecb697ff.md [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+   - Update the balance.   
+   - Return `"Successfully deposited $[amount]. New balance: $[balance]"`.
+
+5. If the amount is less than or equal to `0`, the `deposit` method should return `"Deposit amount must be greater than zero."`.
+
+6. You should define a method named `withdraw` that takes an amount as a parameter. This method should update the current balance according to withdrawals. When the amount to be withdrawn is greater than `0` and less than or equal to the current balance, it should:
+
+   - Push a new object to the transactions array with a type of withdraw and the amount withdrawn.
```
Comment (moT01) [inline curriculum/challenges/english/25-front-end-development/lab-bank-account-manager/6718d2d59337c822ecb697ff.md:30]:
```suggestion
   - Push a new object to the transactions array with a type of `withdraw` and the amount withdrawn.
```

File: curriculum/challenges/english/25-front-end-development/lab-bank-account-manager/6718d2d59337c822ecb697ff.md [hunk 12]
```diff
@@ -0,0 +1,275 @@
...
+   - Update the balance.
+   - Return `"Successfully withdrew $[amount]. New balance: $[balance]"`. 
+
+7. If the amount to be withdrawn is less than or equal to `0` or greater than the current balance, the `withdraw` method should return `"Insufficient balance or invalid amount."`.
+
+
+8. You should define a method named `checkBalance` that returns the current balance in the format `"Current balance: $[balance]"`.
+
+9. You should define a method named `listAllDeposits` that iterates through the transactions array and returns all deposits in the format `"Deposits: deposit,deposit,..."`.
```
Comment (moT01) [inline curriculum/challenges/english/25-front-end-development/lab-bank-account-manager/6718d2d59337c822ecb697ff.md:39]:
```suggestion
9. You should define a method named `listAllDeposits` that iterates through the transactions array and returns all deposits in the format `"Deposits: amount,amount,..."`.
```

I think this makes it a little more clear.

File: curriculum/challenges/english/25-front-end-development/lab-bank-account-manager/6718d2d59337c822ecb697ff.md [hunk 13]
```diff
@@ -0,0 +1,275 @@
...
+
+10. You should define a method named `listAllWithdrawals` that iterates through the transactions array and returns all withdrawals in the format `"Withdrawals: withdrawal,withdrawal,..."`.
```
Comment (moT01) [inline curriculum/challenges/english/25-front-end-development/lab-bank-account-manager/6718d2d59337c822ecb697ff.md:41]:
```suggestion
10. You should define a method named `listAllWithdrawals` that iterates through the transactions array and returns all withdrawals in the format `"Withdrawals: amount,amount,..."`.
```

same

**Review by moT01 — CHANGES_REQUESTED** [review]
This was fun @zairahira 🎉 I left a few comments to try and clarify some of the user stories. Other than that, LGTM 🎉 

One thing that we could do to expand this a little is make a few more user stories like:
"You should create a new instance of `BankAccount` named `myAccount`."
"Your `myAccount` bank account should have at least five transactions."
"Your `myAccount` bank account should have at least two deposits."
"Your `myAccount` bank account should have at least two withdrawals."
"Your `myAccount` bank account should have a balance greater than $100."

Something like that. They would be pretty easy to write and write tests for - feel free to add something like that if you want. Otherwise, I'm fine with getting this in without that.

Actually, now that a take a second look - the `updateBalance` function doesn't really seem like it's needed or something. Here's the code I wrote based on the user story:

<details><summary>code</summary>

```js
updateBalance({type, amount}) {
  type === 'deposit' ? this.balance += amount : this.balance -= amount;
}
```

</details>

It doesn't explain that I am supposed to go through all the transactions and update the balance (found out by looking at your solution). I guess I was a little confused. Here's the code I used to pass:

<details><summary>code</summary>

```js
class BankAccount {
  constructor() {
    this.balance = 0;
    this.transactions = [];
  }

  updateBalance({type, amount}) {
    type === 'deposit' ? this.balance += amount : this.balance -= amount;
  }

  deposit(amount) {
    if(amount > 0) {
      this.balance += amount;
      this.transactions.push({ type: 'deposit', amount })
      return `Successfully deposited $${amount}. New balance: $${this.balance}`;
    } else {
      return "Deposit amount must be greater than zero."
    }
  }

  withdraw(amount) {
    if(amount > 0 && this.balance >= amount) {
      this.balance -= amount;
      this.transactions.push({ type: 'withdraw', amount })
      return `Successfully withdrew $${amount}. New balance: $${this.balance}`;
    } else {
      return "Insufficient balance or invalid amount."
    }
  }

  checkBalance() {
    return `Current balance: $${this.balance}`
  }

  listAllDeposits() {
    const deposits = this.transactions.filter(t => t.type === 'deposit');
    return `Deposits: ${deposits.map(d => d.amount).join(',')}`
  }

  listAllWithdrawals() {
    const withdrawals = this.transactions.filter(t => t.type === 'withdraw');
    return `Withdrawals: ${withdrawals.map(w => w.amount).join(',')}`
  }
}
```

</details>

Note that I just use `this.balance +=/-= amount` in the `deposit`/`withdraw` methods - so I didn't use the `updateBalance` function. Now that I read the user stories again ("Update the balance"), I see that that's what you intended me to do. I think I would vote to just remove that user story and the tests for it - but we could keep it as well. I would clarify some of those user stories if we do - maybe...

User Story 3:

```
Inside the `BankAccount` class, you should define a method named
`updateBalance` that goes through the `transactions` array and updates
the balance according to the type of transaction. If the transaction type is
`deposit`, it should add the amount to the balance. If the transaction type
is `withdraw`, it should subtract the amount from the balance.
```

Or something like that anyway. Then in the withdraw and deposit user story lists. ``Run the `updateBalance` method to update the balance.`` instead of `Update the balance.`

**Comment by zairahira** [comment]
Thank you @Dario-DC and @moT01 for reviewing the lab in detail and sharing the suggestions.

I have made the following changes:

- Removed the `updateBalance` method entirely along with user stories and tests.
- Extended the user stories and tests.

Let me know if these changes are good to go.

**Review by moT01 — APPROVED** [review]
LGTM 🎉

**Pull request merged**